### ISSUE TITLE
StreamReader.ReadLine overload returning string with line ending

### ISSUE THREAD

**Issue by melanchall** [dotnet/runtime#32128]
I suggest to create `StreamReader.ReadLine` overload like this:

```csharp
ReadLine(bool includeLineEnding)
```

**Task that led to the feature request**

I'm writing CSV reader which reads data from file line by line using `StreamReader.ReadLine`. But sometimes a string value in CSV data includes `\r`, `\n` or both. In this case I read next line and merge two lines with `Environment.NewLine`. But it's obviously wrong. For example, on Windows I'll get `\r\n` in a value, but what if only `\n` was in the string value?

Unfortunately `ReadLine` doesn't include exact line ending so there is no way to know what it was.

**Questions on the web**

I suppose this feature is needed not only to me since there are similar threads on the web:

* https://stackoverflow.com/questions/14978095/stream-reader-readline-detect-newline-characters
* https://social.msdn.microsoft.com/Forums/vstudio/en-US/27f78a6f-4ce6-406a-a393-3b97b2bdf90b/read-lines-including-line-terminator?forum=csharpgeneral
* https://www.reddit.com/r/csharp/comments/7lhn8x/streamreader_read_line_with_new_line_character/

**Comment by melanchall** [comment]
@JeremyKuhne Returning line ending is OK too, but I suppose it's not a problem to trim a line if someone wants to get string without `\n` and `\r`. For me it would be better just to return string with all line endings.

Right now my workaround is manual reading via buffer and `ReadBlock`.

**Comment by JeremyKuhne** [comment]
@melanchall The reason I'm asking is that either way can force extra avoidable allocations on users if they end up needing to append or trim. Getting a good grip on potential use scenarios will help clarify which variant we would add or if we would potentially want both.

**Comment by melanchall** [comment]
@JeremyKuhne Makes sense. As for me I need line ending attached to line, I don't trim lines. At now, as I said, I've implemented desired behavior via manual reading with `StreamReader.ReadBlock` method.

But my task is not performance-critical, `ReadLine(out string lineEnding)` will good too, I'll just append line ending to text.

**Comment by mariusGundersen** [comment]
I want this feature too, for a specific reason: It's not possible to tell if the stream ended with a newline or not. 

So for example it's not possible to tell if a file ends with a newline or not, as the ReadLine method returns whatever is left of the stream. It could be that the stream ends with a newline character, like many text files do, but it could also end without a newline character, and there is no way to tell from ReadLine which one it is.

Therefore I think it should return one of four options:
* `\r`
* `\n`
* `\r\n`
* `` (empty string if the stream ends without a newline)

In my specific scenario I make a range request to a very large text file, and the range ends at an arbitrary point, and I need to know if that arbitrary point is at the end of a line or not, so I can know if I have a whole line or not.

**Comment by mariusGundersen** [comment]
Just want to share this very simple solution that unfortunately isn't very efficient:

```csharp
    public sealed class LineReader : IDisposable
    {
        private readonly StreamReader _streamReader;

        public LineReader(Stream stream)
        {
            _streamReader = new StreamReader(stream);
        }

        public void Dispose() => _streamReader.Dispose();

        public string ReadLine(out string newline)
        {
            var sb = new StringBuilder();

            while (true)
            {
                var b = _streamReader.Read();
                switch (b)
                {
                    case -1:
                        newline = "";
                        return sb.ToString();

                    case '\r' when _streamReader.Peek() == '\n':
                        _streamReader.Read();
                        newline = "\r\n";
                        return sb.ToString();

                    case '\r':
                        newline = "\r";
                        return sb.ToString();

                    case '\n':
                        newline = "\n";
                        return sb.ToString();

                    default:
                        sb.Append((char)b);
                        break;
                }
            }
        }
    }
```

Making a more efficient implementation, for example one that is async, would require reimplementing most of StreamReader.
